cocotb/share/lib/gpi/GpiCommon.cpp

```diff
@@ -122,15 +125,39 @@ int gpi_register_impl(GpiImplInterface *func_tbl)
     return 0;
 }
 
-void gpi_embed_init(gpi_sim_info_t *info)
+void gpi_register_sim_event_callback(void (*func)(gpi_event_t, const char *))
 {
-    if (embed_sim_init(info))
-        gpi_embed_end();
+    sim_event_callbacks.push_back(func);
+}
+
+void gpi_register_sim_end_callback(void (*func)(void))
+{
+    sim_end_callbacks.push_back(func);
+}
+
+void gpi_register_extension(const char* name, int (*func)(int, char const* const*))
+{
+    extension_entry_functions.emplace(name, func);
+}
+
+void gpi_embed_init(int argc, char const* const* argv)
+{
+    for (auto& elem : extension_entry_functions) {
+        int ret;
+        auto name = elem.first;
+        auto func = elem.second;
+        if (ret = func(argc, argv)) {
+            LOG_ERROR("Error while loading extension '%s'. Return code was '%d'.", name.c_str(), ret);
+            gpi_embed_end();
+            break;
+        }
+    }
+    extension_entry_functions.clear();
 }
 
 void gpi_embed_end()
 {
-    embed_sim_event(SIM_FAIL, "Simulator shutdown prematurely");
+    gpi_embed_event(SIM_FAIL, "Simulator shutdown prematurely");
```

Comment:
Our chance to change this to
```suggestion
    gpi_embed_event(SIM_FAIL, "Simulator shut down prematurely");
```
? (Feels a bit stupid to mention these tiny things, but our code is in such a good shape these days...)